iterator.go

```diff
@@ -119,69 +114,133 @@ func (r RootLookuper) ExchangeWithServer(ctx context.Context, m *dns.Msg,
 }
 
 // Iterate is an iterative lookup implementation
-// revive:disable:cognitive-complexity
-// revive:disable:cyclomatic
 func (r RootLookuper) Iterate(ctx context.Context, name string,
 	qtype uint16, startAt string,
 ) (*dns.Msg, error) {
-	// revive:enable:cognitive-complexity
-	// revive:enable:cyclomatic
-	if startAt == "" {
-		startAt = pickRoot() + ":53"
+	req := new(dns.Msg)
+	req.SetQuestion(dns.Fqdn(name), qtype)
+	return r.IterateMsg(ctx, req, startAt)
+}
+
+// IterateMsg is an iterative exchange implementation
+func (r RootLookuper) IterateMsg(ctx context.Context, req *dns.Msg,
+	startAt string,
+) (*dns.Msg, error) {
+	var server string
+
+	switch {
+	case startAt != "":
+		server = startAt
+	case r.Start != "":
+		server = r.Start + ":53"
+	default:
+		server = pickRoot() + ":53"
+	}
+
+	msg, err := r.sanitizeRequest(req)
+	if err != nil {
+		return nil, err
 	}
-	server := startAt
-	name = dns.Fqdn(name)
-	msg := r.newMsgFromParts(name, qtype)
-	resp, err := r.ExchangeWithServer(ctx, msg, server)
+
+	return r.doIterate(ctx, msg, server)
+}
+
+func (RootLookuper) sanitizeRequest(req *dns.Msg) (*dns.Msg, error) {
+	if req == nil || len(req.Question) != 1 {
+		return nil, errors.ErrBadRequest()
+	}
+
+	// assemble new equivalent request
+	msg := &dns.Msg{
+		MsgHdr: dns.MsgHdr{
+			Id:               dns.Id(),
+			RecursionDesired: false,
+		},
+		Question: req.Question,
+	}
+
+	msg = msg.SetEdns0(2048, false)
```

Comment:
This is bad for two reasons:
1. Hard coded 2048 bit (it should be `DefaultUDPSize`, and/or define a `MaxUDPSize` (Why we do not just use `dns.DefaultMsgSize` and `dns.MaxMsgSize` ?)

2. It is abusing EDNS0, if we just want to allow a larger message size (and we do not fiddle with EDNS) we could do something like:
```go 
o := new(dns.OPT)
o.SetUDPSize(dns.DefaultMsgSize) // This is actually 4096
msg.Extra = append(msg.Extra, o)
```

Reply:
I copied that line pristine from the current implementation of iterator.go... I don't know the reasoning behind

Reply:
@karasz  as this came straight from RootLookuper.newMsgFromParts(), should this be fixed there first? how exactly?

Reply:
> I copied that line pristine from the current implementation of iterator.go... I don't know the reasoning behind

that does not mean that it is the right way.

Reply:
> @karasz as this came straight from `RootLookuper.newMsgFromParts()`, should this be fixed there first? how exactly?

yes, it needs to be fixed there, the least we should do is replace there `2048` with `dns.DefaultMsgSize`

Reply:
> > @karasz as this came straight from `RootLookuper.newMsgFromParts()`, should this be fixed there first? how exactly?
> 
> yes, it needs to be fixed there, the least we should do is replace there `2048` with `dns.DefaultMsgSize`

done as a new commit because the issue wasn't introduced by the refactoring commit